### ISSUE TITLE
sequenced vs. separate tactics with tactic-in-term can cause unexpected typeclass resolution

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5143]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5143
Date: 2016-10-15 01:38:00 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;
To:   

Last updated: 2016-10-15 19:46:46 +0200

## Bugzilla Comment ID: 14363
Date: 2016-10-15 01:38:00 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;

This case is very similar to
https://coq.inria.fr/bugs/show_bug.cgi?id=5123
but exists even with the fix for 5123 (using 8.6 version 05ad4f4):

However, instead of unshelve leading to an unexpected case of typeclass resolution, in this case it is sequencing tactics with ";" vs. ".", such that the second tactic contains a tactic-in-term.

In the example below, the first Goal shows the bug, and the second two show that (1) the bug doesn't occur without tactic-in-term and (2) the bug doesn't occur when the tactic sequence is "."-delimited instead of ";"

Class Eqdec A := eqdec : forall a b : A, {a=b}+{a<>b}.

Inductive vect A : nat -> Type :=
| vnil : vect A 0
| vcons : forall (h:A) (n:nat), vect A n -> vect A (S n).

Context `{vect_sigT_eqdec : forall A : Type, Eqdec A -> Eqdec {a : nat & vect A a}}.

Ltac uscore term :=
  match goal with
  | _ => uscore open_constr:(term _) (*fails if there are too many _'s*)
  | _ => fun tac => tac term
  end.

Tactic Notation "upose" open_constr(term) "as" ident(H) :=
  uscore term ltac:(fun x => pose x as H).
  
Typeclasses eauto := debug #1.

Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H;
  lazymatch type of H with
      | {?X = _}+{_} => unify X ltac:(eapply existT; exact a)
  end. (*Does typeclass resolution - note debug output*)
  Unshelve.
  Fail all:let N := numgoals in guard N = 3.
Abort.

Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H;
  lazymatch type of H with
      | {?X = _}+{_} => unify X (existT (vect A) n a)
  end. (*Does no typeclass resolution*)
  Unshelve.
  all:let N := numgoals in guard N = 3.
Abort.

Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H.
  lazymatch type of H with
      | {?X = _}+{_} => unify X ltac:(eapply existT; exact a)
  end. (*Does no typeclass resolution*)
  Unshelve.
  all:let N := numgoals in guard N = 3.
Abort.

## Bugzilla Comment ID: 14364
Date: 2016-10-15 03:03:04 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;

It's not the fault of the tactic-in-term, as this shows:

Typeclasses eauto := debug #1.
Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H;
  let o := open_constr:(_) in 
  assert o; [eapply existT|].

Here, the eapply in the asserted subgoal sets off typeclass resolution, as noted by the debug output.  I guess the commonality in the buggy cases is that an evar is instantiated in the tactic sequence, and that somehow causes inappropriate typeclass resolution in other evars created in that sequence.

## Bugzilla Comment ID: 14365
Date: 2016-10-15 03:17:17 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;

Also note that refine, unlike eapply, does not produce the unexpected typeclass resolution.

Typeclasses eauto := debug #1.
Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H;
  let o := open_constr:(_) in 
  assert o; [refine (existT _ _ _)|].

## Bugzilla Comment ID: 14373
Date: 2016-10-15 19:46:46 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;

Not quite - using refine merely delays the problem:

Goal forall A n (a b : vect A n), {a=b}+{a<>b}.
Proof.
  intros A n a b.
  upose vect_sigT_eqdec as H;
  lazymatch type of H with
      | {?X = _}+{_} => unify X ltac:(refine (existT _ _ _); exact a)
  end.
  Fail left.

with the error being:

The command has indeed failed with message:
Unable to satisfy the following constraints:
In environment:
A : Type
n : nat
a, b : vect A n

?e : "Eqdec A"
?s : "{a : nat & vect A a}"

In other words, the "left" tactic was enough to trigger typeclass resolution!
